homeassistant/components/ihc/__init__.py

```diff
@@ -2,70 +2,110 @@
+            supported. See https://www.home-assistant.io/integrations/ihc/
+            """
+        )
+        migrate_configuration(hass)
+        return False
     return True
 
 
-def ihc_setup(
-    hass: HomeAssistant,
-    config: ConfigType,
-    controller_conf: ConfigType,
-    controller_index: int,
-) -> bool:
-    """Set up the IHC integration."""
-    url = controller_conf[CONF_URL]
-    username = controller_conf[CONF_USERNAME]
-    password = controller_conf[CONF_PASSWORD]
-
-    ihc_controller = IHCController(url, username, password)
-    if not ihc_controller.authenticate():
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the IHC Controller from a config entry."""
+    controller_id: str = str(entry.unique_id)
+    url: str = entry.data[CONF_URL]
+    username: str = entry.data[CONF_USERNAME]
+    password: str = entry.data[CONF_PASSWORD]
+    autosetup: bool = entry.data[CONF_AUTOSETUP]
+    ihc_controller: IHCController = IHCController(url, username, password)
+    if not await hass.async_add_executor_job(ihc_controller.authenticate):
         _LOGGER.error("Unable to authenticate on IHC controller")
         return False
-    controller_id: str = ihc_controller.client.get_system_info()["serial_number"]
-    # Store controller configuration
     hass.data.setdefault(DOMAIN, {})
     hass.data[DOMAIN][controller_id] = {
         IHC_CONTROLLER: ihc_controller,
-        CONF_INFO: controller_conf[CONF_INFO],
-        IHC_CONTROLLER_INDEX: controller_index,
     }
-    if controller_conf[CONF_AUTOSETUP] and not autosetup_ihc_products(
-        hass, config, ihc_controller, controller_id
-    ):
+    if not await setup_controller_device(hass, ihc_controller, entry):
         return False
-    get_manual_configuration(hass, config, controller_conf, controller_id)
-    # We only want to register the service functions once for the first controller
-    if controller_index == 0:
+    if autosetup:
+        await hass.async_add_executor_job(
+            autosetup_ihc_products, hass, ihc_controller, controller_id
+        )
+    await hass.async_add_executor_job(manual_setup, hass, controller_id)
+    for component in IHC_PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+    entry.add_update_listener(async_update_options)
+    # We only wan to register service functions once, in case you have multiple controllers
+    if len(hass.data[DOMAIN]) == 1:
         setup_service_functions(hass)
     return True
+
+
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        config_entry, IHC_PLATFORMS
+    )
+    if not unload_ok:
+        return False
+    controller_id = config_entry.unique_id
+    ihc_controller = hass.data[DOMAIN][controller_id][IHC_CONTROLLER]
```

Comment:
I use the controller id because if you have more than one controller, you need to specify the controller id (controller serial), when using the ihc services. This make it easier/faster to look it up.

Reply:
Our common design pattern is to store instance specific data with the config entry id as the key, that way its easy recognizable. So it would be great if you could change it.

Reply:
I have changed it to use the entry id